### ISSUE TITLE
Add support for using AWS Secrets as a credential store

### ISSUE THREAD

**Issue by darren-reddick** [devopsgoat/terraform-aws-transfer#3]

**Comment by darren-reddick** [comment]
Enhancement provided by commit 952002ba0dad198faf63a92f5c9868a7a6a8bb09

**Comment by raolivei** [comment]
Hi @devopsgoat 
Thanks for sharing this enhancement!

I have successfully deployed the recent changes and created a SecretsManager secret in the following format:

Encryption key
DefaultEncryptionKey

Secret name
SFTP/test

Secret ARN
arn:aws:secretsmanager:us-east-1: <account-id>: secret:SFTP/test-DViWYg

Secret Key | Secret Value
-- | --
UserId | test
Role | arn:aws:iam::<account-id>:role/transfer-user-iam-role
Password | test!
HomeDirectoryDetails | [{\"Entry\": \"/\", \"Target\": \"/scx-sftp/${Transfer:UserName}\"}]

However, I am getting a permission denied when trying to SFTP – same error message I get when entering an invalid username so I am suspecting that Lambda is either not querying SecretsManager values, or the secret format should be different (something like plain text instead of key,value mapping?)

I will keep digging and will post here if I find the solution.

**Comment by darren-reddick** [comment]
Hi,
To help determine where the issue lies try these steps first:

* Run the "Test" function in the AWS Transfer console using those credentials. This will show if a response is coming back from the API Gateway
* Check the logs for the Lambda execution

**Comment by raolivei** [comment]
Here it goes.

Based on transfer-server test's response message, I would say that it might be related to HomeDirectoryDetails value's format. But I am not sure which format is SecretManager expecting.

From transfer-server test (TestIdentityProvider Response):

```JSON
{

"Response": "{\"Role\": \"arn:aws:iam::<<account_id>>:role/transfer-user-iam-role\",\"HomeDirectoryType\": \"LOGICAL\",\"HomeDirectoryDetails\": \"[{\\\"Entry\\\": \\\"/\\\", \\\"Target\\\": \\\"/scx-sftp/${Transfer:UserName}\\\"}]\"}",

"StatusCode": 200,

"Message": "HomeDirectoryDetails is not a valid JSON document",

"Url": "https://<<server_id>>.execute-api.us-east-1.amazonaws.com/prod/servers/s-3e02bcab4c0946e8a/users/test/config"

}
```

From Lambda logs in CloudWatch:

```
17:54:17
START RequestId: 755ce9a3-2fb7-49b5-9262-8b307ac63d6e Version: $LATEST
17:54:17
Username: test, ServerId: s-3e02bcab4c0946e8a
17:54:17
Secrets Manager Region: us-east-1
17:54:17
{"UserId":"test","Role":"arn:aws:iam::<<account_id>>:role/transfer-user-iam-role","Password":"test!","HomeDirectoryDetails":"\"[{\\\"Entry\\\": \\\"/\\\", \\\"Target\\\": \\\"/scx-sftp/${Transfer:UserName}\\\"}]\""}
17:54:17
Found Secret String
17:54:17
HomeDirectoryDetails found - Applying setting for virtual folders
17:54:17
Completed Response Data: {"Role": "arn:aws:iam::<<account_id>>:role/transfer-user-iam-role", "HomeDirectoryDetails": "\"[{\\\"Entry\\\": \\\"/\\\", \\\"Target\\\": \\\"/scx-sftp/${Transfer:UserName}\\\"}]\"", "HomeDirectoryType": "LOGICAL"}
17:54:17
END RequestId: 755ce9a3-2fb7-49b5-9262-8b307ac63d6e
17:54:17
REPORT RequestId: 755ce9a3-2fb7-49b5-9262-8b307ac63d6e Duration: 665.85 ms Billed Duration: 700 ms Memory Size: 128 MB Max Memory Used: 93 MB
```

**Comment by darren-reddick** [comment]
Hi,
Luckily I have been working on using Secrets today! :0)
I have just updated the README for the secrets example with some sample tf code for creating the secret as an example of how the secret should look. I havent tested the full example so it is in the "examplesecret" branch for the moment.

There was also a minor fix to the IAM perms in the example to allow proper access to the bucket.

**Comment by raolivei** [comment]
Hi @devopsgoat ,

I confirm that your fix to the *HomeDirectoryDetails* on secrets manager works! I am able to successfully connect to the user's home dir.

[{\"Entry\": \"/\", \"Target\": \"/scx-sftp/${Transfer:UserName}\"}]

**Comment by darren-reddick** [comment]
Confirmed as fixed in 6d331f032c07ab1179a8a1efa9de0ef6ebf641a9
